Update RESTful API (#1997)
pytorch/fairseq: examples/simultaneous_translation/docs/server_api.md

## **Server RESTful API**

Assuming the server address is *server_url*

---

## Get number of sentences in test set

* **Method:** : `GET`
  
*  **Endpoint** : *server_url*
  
*  **URL Params:**
 
* **Success Response:**
    
    ```
    {
        "num_sentences": N
    }
    ``` 

Where `N` is the number of sentences in test set

---
## Start a new evaluation session
* **Method:** : `POST`
  
*  **Endpoint** : *server_url*
  
*  **URL Params:** 

Should be used everytime starting to evaluate a new model.

___
## Obtain a source word from server

* **Method:** : `GET`
  
*  **Endpoint** : *server_url*/src
  
*  **URL Params:** 
  
    ```
    {
        "sent_id": i
    }
    ``` 
 
* **Success Response:**
    
    ```
    {
        "sent_id": i,
        "segment_id": segment_id,
        "segment": word
    }
    ```

A source word in sentence *i* will be received from server. 
Notice that the word is not tokenized. 
```segment_id``` is the position of the word in sentence, starting from zero.
The server will automatically move the pointer to next word. 
When the pointer reach the end of sentence, "<\/s>" will be sent. 
For example, for sentence "A B C". 

> The first time the request happens, "A" will be received,  the second time will be "B", the three time will be "C",  and after then will be always "<\/s>". The request should be made everytime the model decides to read a new word.

Reviewer: ```suggestion
The first time the request happens, "A" will be received,  the second time will be "B", the third time will be "C",  and after that, it will be always "<\/s>". The request should be made everytime the model decides to read a new word.
```

---
## Obtain a segment of speech from server

* **Method:** : `GET`
  
*  **Endpoint** : *server_url*/src
  
*  **URL Params:** 
  
    ```
    {
        "sent_id": i,
        "segment_size": t_ms
    }
    ``` 
 
* **Success Response:**
    
    ```
    {
        "sent_id": i,
        "segment_id": segment_id,
        "segment": list_of_numbers
    }
    ```

> The wav form of a segement speech utterence will be received in the format of list of numbers.

Reviewer: ```suggestion
The wav form for the segment of a speech utterance will be received in the format of list of numbers.
```

The sample rate on the server is 16000Hz. By default, the length of list is 160, or 10 ms in time.

> There is optional ```segment_size``` parameter in the request, the unit is ms. A customized length of segment can be requested. However, the length of segment in time can only be the multiple of 10ms. It would be rounded if not. For example, if ```t_ms = 301``` is requested, the returned utterence will be 300ms long.

Reviewer: ```suggestion
There is an optional `segment_size` parameter in the request, the unit is ms. A customized length of segment can be requested. However, the length of a segment in time can only be the a multiple of 10ms. It will be rounded if not. For example, if `t_ms = 301` is requested, the returned segment will be 300ms long.
```

Again, the request should be made every time the model decides to read a new segment of speech utterence. 

---

> ## Send a translated token to server

Reviewer: ```suggestion
## Send a translated token to the server
```

* **Method:** : `PUT`
  
*  **Endpoint** : *server_url*/hypo
  
*  **URL Params:** 
  
    ```
    {
        "sent_id": i,
    }
    ``` 
*  **Body** (raw text)

    ```
    WORD
    ```

> After the token is sent, the server will record the delays (length of source context) the model use to predict the token. Notice that the content should be detokenized. If there is space in ```WORD```, it will be considered as multiple words splitted by space. Once finishing the sentence, an end of sentence token "<\/s>" should be sent to the server.

Reviewer: ```suggestion
After the token is sent, the server will record the delay (length of source context) the model used to predict the token. Notice that the content should be detokenized. If there is a space in `WORD`, it will be considered as multiple words split by space. In order to end a translation hypothesis, an end of sentence token "<\/s>" should be sent to the server.
```

---

> ## Get evaluation scores from server

Reviewer: ```suggestion
## Get evaluation scores from the server
```

* **Method:** : `GET`
  
*  **Endpoint** : *server_url*/result
  
*  **URL Params:**
 
* **Success Response:**
  ```
    {
        "BLEU": BLEU, 
        "TER": TER, 
        "METEOR": METEOR,
        "DAL": DAL,
        "AL": AL,
        "AP": AP
    }
    ```

> Make sure that make the request after finishing all the translation.

Reviewer: ```suggestion
Make sure to make this request after finishing all the translations.
```

---

> ## **The structure of evaluation client**

Reviewer: ```suggestion
## **The structure of the evaluation client**
```

> Here is example pseudocode of client. 

Reviewer: ```suggestion
Here is example pseudocode for a client. 
```
